[GOBBLIN-2107] Delete adhoc flowSpecs from flowCatalog bug fix

umustafi opened the pull request:
* if job compilation fails then ensure quota is released and flow compilation event sent in all cases
	* fix bug where adhoc flows are not deleted in multi-active scheduler case by re-adding deletion to DagManager.addDag()

Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!


### JIRA
- [X] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-2107


### Description
- [X] Here are some details about my PR, including screenshots (if applicable):
Delete adhoc `flowSpecs` from `flowCatalog` to avoid build up of adhoc flowSpecs in catalog even when job compilation fails by adding the deletion in a finally block in non-MA scheduler case. 

1. Previous PR: https://github.com/apache/gobblin/pull/3944/files causes a bug where adhoc flows are not deleted when multi-active scheduler is enabled. This PR re-adds deletion of `flowSpec` to the `dagManager` to be called by the active host for the MA scheduler case (there will only be one active host if `dagManager` is enabled otherwise code will use `dagProcessingEngine`). 
2. It also ensures quota is released and failed flow compilation event sent by the `FlowCompilationValidationHelper` in all compilation failure cases. This was being incorrectly handled before. 

### Tests
- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

- added a unit test `testDeleteFlowSpecCalledForMultiActivePath` that validates that the `compileAndSubmitFlowToDagManager` used by the `DagActionStoreChangeMonitor` attempts deleting adhoc flowSpecs from the catalog. Note that this test does not definitively ensure that this method is called by the `DagActionStoreChangeMonitor` which proves more complex but verifies if that method is used then the flow will be deleted after compilation. 

### Commits
- [X] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    3. Subject is limited to 50 characters
    4. Subject does not end with a period
    5. Subject uses the imperative mood ("add", not "adding")
    6. Body wraps at 72 characters
    7. Body explains "what" and "why", not "how"

phet commented on gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/DagActionStoreChangeMonitor.java:
```diff
@@ -321,8 +321,10 @@ protected void submitFlowToDagManagerHelper(DagActionStore.DagAction dagAction,
     } catch (SpecNotFoundException e) {
-      log.warn("Spec not found for flowId {} due to exception {}", flowId, e.getMessage());
-      launchSubmissionMetricProxy.markFailure();
+      log.info("Spec not found for flowId {} due to deletion by active dagManager host due to exception {}",
+          flowId, e.getMessage());
+      // TODO: mark this failure if there are other valid cases of this exception
+      // launchSubmissionMetricProxy.markFailure();
```
1. let's figure out whether there are other valid cases!

not sure if I'm misunderstanding... but a TODO is far more reasonable when the impl change would be considerable effort, so we plan to come back, vs. when the task is to figure out how our system behaves.  usually, the latter is worth resolving ASAP.

give that, was it incorrect previously to `markFailure()` and we no longer want to do?

2. on the message: `"Spec not found for flowId {} due to deletion by active dagManager host due to exception {}",`

is that certain it was due to deletion by active DM or is that speculative?

umustafi replied:
It's speculative, I think there could be a case where a SpecNotFoundException occurs in a case _not_ due to deletion by active DM if something goes wrong at a different point in the system (mistaken deletion NOT from the deleteFlowSpecIfAdhoc code path). It's hard to definitively say all of these errors are due to an active DM deleting it but I would expect them all to be.

phet commented on gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/Orchestrator.java:
```diff
@@ -289,7 +293,7 @@ public void submitFlowToDagManager(FlowSpec flowSpec, Dag<JobExecutionPlan> jobE
       consumption of a launch type event is committed to the consumer.
        */
-      this.dagManager.addDag(jobExecutionPlanDag, true, true);
+      this.dagManager.addDagAndRemoveAdhocFlowSpec(flowSpec, jobExecutionPlanDag, true, true);
```
now the attempt to remove an ad hoc spec would occur twice, since it's also on line 267.

I believe your aim is to have removal also happen in the path from [DagActionStoreChangeMonitor](https://github.com/apache/gobblin/blob/4721d086f78553cab28e46bf406787df6cc2dea8/gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/DagActionStoreChangeMonitor.java#L318), correct?

couldn't that run into the same issue, where if the flow compilation validation helper fails (see L281 directly above), then the call to this new combo addDag+removeAdHoc would be skipped?

this new PR helps me realize I had missed the `DagActionStoreChangeMonitor` case in my recent PR.  the crux of that one was to divide adding the dag from removing the ad hoc spec.  the latter MUST happen, suggesting it belongs inside a `finally`, whereas the former is only best-effort.

phet commented on gobblin-service/src/main/java/org/apache/gobblin/service/modules/utils/FlowCompilationValidationHelper.java:
```diff
@@ -140,31 +140,39 @@ public Optional<Dag<JobExecutionPlan>> validateAndHandleConcurrentExecution(Conf
+
+    try {
+      if (jobExecutionPlanDag == null || jobExecutionPlanDag.isEmpty()) {
+        optionalJobExecutionPlan = Optional.absent();
```
this logic gets convoluted.  why not return the `Optional.absent()` as previously and leave it to a `finally` to clean up?

phet commented on gobblin-service/src/main/java/org/apache/gobblin/service/modules/utils/FlowCompilationValidationHelper.java:
```diff
@@ -140,31 +140,39 @@ public Optional<Dag<JobExecutionPlan>> validateAndHandleConcurrentExecution(Conf
+        optionalJobExecutionPlan = Optional.absent();
+      }
+    } finally {
+      if (optionalJobExecutionPlan == null || !optionalJobExecutionPlan.isPresent()) {
```
again, convoluted to need to handle a case where a var may not yet have been initialized (even ironic, since the point of `Optional` is that we no longer need `null` checks)

where are the test cases to validate correctness?

umustafi replied:
updated this code path to make it more clear. I realized the quota manager does not need to be updated when the compilation fails so this looks much simpler now

phet reviewed (commented):
[I pointed out](https://github.com/apache/gobblin/pull/3944/files#r1591517872) in my earlier PR (you linked to) that we're missing test cases.  I had added ones relating to my change.  what would it take to exercise any of the bug fixes that motivate these latest mods?

umustafi commented:
> [I pointed out](https://github.com/apache/gobblin/pull/3944/files#r1591517872) in my earlier PR (you linked to) that we're missing test cases. I had added ones relating to my change. what would it take to exercise any of the bug fixes that motivate these latest mods?

I added a unit test `testDeleteFlowSpecCalledForMultiActivePath` that validates that the `compileAndSubmitFlowToDagManager` used by the `DagActionStoreChangeMonitor` attempts deleting adhoc `flowSpecs` from the catalog. Note that this test does not definitively ensure that this method is called by the `DagActionStoreChangeMonitor` which proves more complex but verifies if that method is used then the flow will be deleted after compilation.

phet reviewed (approved):
nice re-work!

Will-Lo reviewed (commented):
Are we not including this function in the DagProcEngine as well? Or is that out of scope

umustafi commented:
> Are we not including this function in the DagProcEngine as well? Or is that out of scope

Good call out. I'm going to put that as out of scope, @arjun4084346 or I will add this in separate PR for `DagProcessingEngine`
